zk/legacy_executor_verifier/legacy_executor_verifier.go

```diff
@@ -89,150 +86,65 @@ func NewLegacyExecutorVerifier(
-		requestsMap:      make(map[uint64]uint64),
-		responses:        make([]*VerifierResponse, 0),
-		responseMutex:    &sync.Mutex{},
 		quit:             make(chan struct{}),
 		streamServer:     streamServer,
 		stream:           stream,
 		witnessGenerator: witnessGenerator,
 		l1Syncer:         l1Syncer,
+		promises:         make([]*Promise[*VerifierResponse], 0),
+		promisesLock:     sync.Mutex{},
+		addedBatches:     make(map[uint64]struct{}),
 	}
 
 	return verifier
 }
 
-func (v *LegacyExecutorVerifier) StopWork() {
-	close(v.quit)
-}
-
-func (v *LegacyExecutorVerifier) StartWork() {
-	go func() {
-		tick := time.NewTicker(1 * time.Second)
-	LOOP:
-		for {
-			select {
-			case <-v.quit:
-				break LOOP
-			case <-tick.C:
-				go v.processOpenRequests()
-			}
-		}
-	}()
-}
-
-func (v *LegacyExecutorVerifier) processOpenRequests() {
-	v.working.Lock()
-	defer v.working.Unlock()
-
-	if len(v.openRequests) == 0 {
-		return
-	}
-
-	// sort the requests by batch number ascending
-	sort.Slice(v.openRequests, func(i, j int) bool {
-		return v.openRequests[i].BatchNumber < v.openRequests[j].BatchNumber
-	})
-
-	// we want to trim down the requests once we have worked through them so keep track of
-	// where we got to
-	successCount := 0
-
-	// process the requests
-	for _, request := range v.openRequests {
-		processed, err := v.handleRequest(context.Background(), request)
-		if err != nil {
-			log.Error("[Verifier] error handling request", "batch", request.BatchNumber, "err", err)
-			break
-		}
-		if !processed {
-			// likely the underlying stage loop is still running so the batch had no transactions
-			// in this case subsequent batches will also be in the same state so exit now and wait
-			// for the next iteration
-			break
-		}
-		successCount++
-	}
-
-	v.openRequests = v.openRequests[successCount:]
-}
-
-func (v *LegacyExecutorVerifier) getNextOnlineExecutor() ILegacyExecutor {
-	var exec ILegacyExecutor
-
-	// attempt to find an executor that is online amongst them all
-	for i := 0; i < len(v.executors); i++ {
-		v.executorNumber++
-		if v.executorNumber >= len(v.executors) {
-			v.executorNumber = 0
-		}
-		temp := v.executors[v.executorNumber]
-		if temp.CheckOnline() {
-			exec = temp
-			break
-		}
-	}
-
-	return exec
-}
-
-func (v *LegacyExecutorVerifier) handleRequest(ctx context.Context, request *VerifierRequest) (bool, error) {
+func (v *LegacyExecutorVerifier) AddRequestUnsafe(ctx context.Context, tx kv.RwTx, request *VerifierRequest) (*Promise[*VerifierResponse], error) {
 	// if we have no executor config then just skip this step and treat everything as OK
 	if len(v.executors) == 0 {
 		response := &VerifierResponse{
 			BatchNumber: request.BatchNumber,
 			Valid:       true,
 		}
-		v.handleResponse(response)
-		return true, nil
+		return &Promise[*VerifierResponse]{
+			result: response,
+			err:    nil,
+		}, nil
 	}
 
-	execer := v.getNextOnlineExecutor()
+	execer := v.getNextOnlineAvailableExecutor()
```

Comment:
this is initialized very very far away from the place where it is used -- maybe let's move it?
or maybe let's even move it into the actual verification func?